// mns-core-ui/src/mns-core-ui/templates/navigation/test/pageObject/navigation.mobile.page.js
import Page from '../../../../../test/pageObject/page';

const USER_COOKIE = 'MS_USER_COOKIE';
const ANIMATION_WAIT_TIMEOUT = 1000;
const DEFAULT_PATH = 'examples/templates/navigation/body';

export default class NavigationMobilePage extends Page {

    constructor(path = DEFAULT_PATH) {
        super(path);
    }

    open() {
        super.open();
        this.mobileViewport();
    }

    logUserIn(value) {
        browser.setCookie({
            name: USER_COOKIE,
            value: value || 'Jon Doe'
        });

        this.open();
    }

    logUserOut() {
        browser.deleteCookie(USER_COOKIE);
        this.open();
    }

    openHamburgerMenu() {
        browser.element('.menu__button--show-level1').click();
        browser.pause(ANIMATION_WAIT_TIMEOUT);
    }

    get hamburgerOpenMenu() {
        return browser.element('.menu__button--show-level1');
    }

    get myAccount() {
        return browser.element('.menu__mobile__account li:nth-child(1)');
    }

    get sparks() {
        return browser.element('.menu__mobile__account li:nth-child(2)');
    }

    get signOut() {
        return browser.element('.menu__mobile__account li:nth-child(3)');
    }

    get register() {
        return browser.element('.menu__mobile__account li:nth-child(4)');
    }
}
