// mns-core-ui/src/mns-core-ui/templates/navigation/js/desktop.js
import UI from '../../../base/js/ui';
import LazyLoad from '../../../base/js/lazyload';

const NAME = 'desktop';

export default class Desktop {

    /**
     * Desktop navigation
     * @param events
     */
    constructor(events) {

        var elements = {
            link: '.nav li > a'
        };

        this.lazyLoad = new LazyLoad(300);
        this.loadImages = this.loadImages.bind(this);

        this.UI = new UI(NAME, elements);
        events && this.UI.analytics(events[NAME]);
    }

    bind() {
        this.lazyLoad.bind();

        this.level1El = document.getElementsByClassName('nav__level1-panel')[0].children;

        for (let iCount = 0; iCount < this.level1El.length; iCount++) {
            this.level1El[iCount].addEventListener('touchstart', this.loadImages);
            this.level1El[iCount].addEventListener('mouseover', this.loadImages);
        }
    }

    unbind() {
        for (let iCount = 0; iCount < this.level1El.length; iCount++) {
            this.level1El[iCount].removeEventListener('touchstart', this.loadImages);
            this.level1El[iCount].removeEventListener('mouseover', this.loadImages);
        }

        this.lazyLoad.unbind();
    }

    loadImages(){
        this.lazyLoad.load();
    }


}
